cephfs: fix unmountVolume function

Rakshith-R opened the pull request:
This commit fixes bug which caused unmountVolume to fail when path was already unmounted.

Fixes: #1860 

### Future concerns ###
~Changes in the following line has direct impact on ability to catch _path not mounted error.~

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented on internal/cephfs/volumemounter.go:
```diff
@@ -258,7 +258,7 @@ func bindMount(ctx context.Context, from, to string, readOnly bool, mntOptions [
 
 func unmountVolume(ctx context.Context, mountPoint string) error {
 	if err := execCommandErr(ctx, "umount", mountPoint); err != nil {
```
This becomes too generic and can cause issues. actual umount error will be captured in the stderr, `execCommandErr
` should also return the stderr or add stderr to the error message.

Madhu-1 reviewed (changes requested):
please add the E2E test.

yati1998 reviewed (commented):
Please change the commit message according to proper guidelines.

humblec commented on internal/cephfs/volumemounter.go:
```diff
@@ -258,7 +258,7 @@ func bindMount(ctx context.Context, from, to string, readOnly bool, mntOptions [
 func unmountVolume(ctx context.Context, mountPoint string) error {
 	if err := execCommandErr(ctx, "umount", mountPoint); err != nil {
-		if strings.Contains(err.Error(), fmt.Sprintf("exit status 32: umount: %s: not mounted", mountPoint)) ||
+		if strings.Contains(err.Error(), fmt.Sprintf("an error (exit status 32) occurred while running umount args: [%s]", mountPoint)) ||
```
@Rakshith-R please check the error string in 2/3 ceph versions from current. Most of the time the assumed error strings has caused the issues or regression so the request.

Madhu-1 replied:
@humblec is this dependent on the ceph version?

nixpanic replied:
no, it should not depend on the Ceph version. The error comes from the `umount` executable, not Ceph.

humblec replied:
ah. .its unmount not mount. ignore

nixpanic commented on e2e/utils.go:
```diff
@@ -28,8 +28,10 @@ const (
 
-	rookTolBoxPodLabel = "app=rook-ceph-tools"
-	rbdmountOptions    = "mountOptions"
+	rookTolBoxPodLabel        = "app=rook-ceph-tools"
```
You might want to send an other PR or add a commit that corrects the name of `ToolBox`. Not required, as you did not introduce the typo :)

nixpanic commented on e2e/cephfs_helper.go:
```diff
@@ -98,6 +98,23 @@ func createCephfsSecret(f *framework.Framework, secretName, userName, userKey st
 }
 
+func unmountCephFSVolume(f *framework.Framework, app *v1.Pod, pvc *v1.PersistentVolumeClaim) error {
+	app, err := f.ClientSet.CoreV1().Pods(f.UniqueName).Get(context.TODO(), app.Name, metav1.GetOptions{})
```
It is not very clean to re-assign variables that were passed as an argument to a function. Using a new variable name is makes the code easier to understand.

Is the getting of the `app` Pod really needed here?

This works on our standard e2e because there is only a single node (minikube). It would be good to add a comment like `FIXME: run the unmount command on the node where the pod is running` or similar. It would give a hint on what is needed to  get addressed when we start running multi-node tests.

Rakshith-R replied:
> It is not very clean to re-assign variables that were passed as an argument to a function. Using a new variable name is makes the code easier to understand.

Okay, app name and volume name is sufficient. I will change the function to use them instead.

> Is the getting of the `app` Pod really needed here?

Need `UID` of `app` pod to form the mount path to run unmount command.
`cmd := fmt.Sprintf("umount /var/lib/kubelet/pods/%s/volumes/kubernetes.io~csi/%s/mount", app.ObjectMeta.UID, pv.ObjectMeta.Name)`

> This works on our standard e2e because there is only a single node (minikube). It would be good to add a comment like `FIXME: run the unmount command on the node where the pod is running` or similar. It would give a hint on what is needed to get addressed when we start running multi-node tests.

I'll do that :+1:

Madhu-1 commented on e2e/cephfs_helper.go:
```diff
@@ -98,6 +98,23 @@ func createCephfsSecret(f *framework.Framework, secretName, userName, userKey st
+	if err != nil {
+		return err
+	}
+	cmd := fmt.Sprintf("umount /var/lib/kubelet/pods/%s/volumes/kubernetes.io~csi/%s/mount", app.ObjectMeta.UID, pv.ObjectMeta.Name)
```
The E2E should be as below

* Get list of running cephfs plugin pods
* Run through all cephfsplugin pods and run `mount |grep path`  and check application is mounted in cephfs plugin pod
* If the mount is present than umount it.

Madhu-1 commented on internal/cephfs/volumemounter.go:
```diff
@@ -257,8 +257,9 @@ func bindMount(ctx context.Context, from, to string, readOnly bool, mntOptions [
-		if strings.Contains(err.Error(), fmt.Sprintf("exit status 32: umount: %s: not mounted", mountPoint)) ||
+	if _, stderr, err := util.ExecCommand(ctx, "umount", mountPoint); err != nil {
+		err = fmt.Errorf("%w stderr: %s", err, stderr)
+		if strings.Contains(err.Error(), fmt.Sprintf("an error (exit status 32) occurred while running umount args: [%s] stderr: umount: %s: not mounted.", mountPoint, mountPoint)) ||
```
Just check for `umount: %s: not mounted`?

Rakshith-R committed:
cephfs: fix unmountVolume function

This commit fixes bug in unmount function which caused
unmountVolume to fail when targetPath was already unmounted.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: small fixes in var name and error msg

rookTolBoxPodLabel to rookToolBoxPodLabel
podPot to podOpt
fix error message for getting PV

Signed-off-by: Rakshith R <rar@redhat.com>

Commit:
e2e: add getDaemonSetLabelSelector() to utils

getDaemonSetLabelSelector returns labels of daemonset given name and
namespace dynamically, needed since labels are not same for helm and
non-helm deployments.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>
Signed-off-by: Rakshith R <rar@redhat.com>

yati1998 commented on e2e/cephfs.go:
```diff
@@ -528,6 +528,23 @@ var _ = Describe("cephfs", func() {
+			By("Create PVC, bind it to an app, unmount volume and check app deletion", func() {
+				pvc, app, err := createPVCAndAppBinding(pvcPath, appPath, f, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to create PVC and application with error %v", err)
```
Use wrapped static errors instead of dynamic. You can refer this PR: https://github.com/humblec/ceph-csi/commit/82b2ad28a6e3b578af701c32bd112e761e9ec602

Rakshith-R committed:
e2e: add listPods()

listPods returns slice of pods matching given ListOptions and
namespace.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add execCommandInDaemonsetPod()

execCommandInDaemonsetPod() executes commands inside given
container of a daemonset pod on a particular node.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add test case for app deletion with unmounted cephFS volume

Checks app deletion when cephFS volume is already unmounted.
Creates app, pvc and binds them. Unmounts the volume through
umount cmd in cephFS plugin and checks app deletion.

Signed-off-by: Rakshith R <rar@redhat.com>

nixpanic commented on e2e/pod.go:
```diff
@@ -121,6 +121,42 @@ func getCommandInPodOpts(f *framework.Framework, c, ns string, opt *metav1.ListO
 	}, nil
 }
 
+func execCommanddInDaemonsetPod(f *framework.Framework, c, daemonsetName, nodeName, containerName, ns string) (string, string, error) {
```
typo: double 'd' in 'Commandd'

nixpanic commented on e2e/pod.go:
```diff
@@ -105,26 +105,30 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
+func listPods(f *framework.Framework, ns string, opt *metav1.ListOptions) ([]v1.Pod, error) {
+	podList, err := f.PodClientNS(ns).List(context.TODO(), *opt)
+	if len(podList.Items) == 0 {
+		return podList.Items, errors.New("podlist is empty")
```
it probably helps debugging if you can add some of the details (namespace, ListOptions) to the error message

nixpanic commented on e2e/pod.go:
```diff
@@ -18,6 +18,21 @@ import (
 	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
 )
 
+func getDaemonSetLabelSelector(f *framework.Framework, ns, dsn string) (string, error) {
```
it is strongly recommended to add comments for new functions, even when they are for e2e/internal usage

nixpanic reviewed (changes requested):
need to fix the typo "Commandd', maybe address other nits

nixpanic reviewed (dismissed):
This looks really clean now, thanks!

Madhu-1 commented:
@Rakshith-R DCO bot is failing

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19
